src/maliput/base/distance_router.cc

```diff
@@ -0,0 +1,217 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#include "maliput/base/distance_router.h"
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "maliput/api/lane.h"
+#include "maliput/api/regions.h"
+#include "maliput/common/maliput_throw.h"
+#include "maliput/routing/derive_lane_s_routes.h"
+#include "maliput/routing/find_lane_sequences.h"
+#include "maliput/routing/phase.h"
+
+namespace maliput {
+namespace {
+
+// Computes the cost of a routing::Phase.
+//
+// Treats all api::LaneSRanges within @p phase the same. Subsequent steps in the
+// DistanceRouter will refine this cost estimate by, for example, considering
+// which api::LaneSRanges will require lane changes to reach the route's end
+// location.
+//
+// @param phase The routing::Phase to compute the cost.
+// @return The minimum length across all api::LaneSRanges in the routing::Phase.
+double ComputeCost(const routing::Phase& phase) {
+  double cost{std::numeric_limits<double>::max()};
+  for (const api::LaneSRange& lane_s_range : phase.lane_s_ranges()) {
+    cost = std::min(cost, lane_s_range.length());
+  }
+  return cost;
+}
+
+// Computes the cost of a routing::Route.
+// @param route The routing::Route to compute the cost.
+// @return The accumulated cost of all the routing::Phases in @p route.
+double ComputeCost(const routing::Route& route) {
+  double cost{0.};
+  for (int i = 0; i < route.size(); ++i) {
+    cost += ComputeCost(route.Get(i));
+  }
+  return cost;
+}
+
+// Determines the maximum cost of all routing::Phases in @p route.
+// @param route The routing::Route to look for the maximum routing::Phase's cost.
+// @return The maximum cost of all routing::Phases in @p route.
+double MaxPhaseCostInRoute(const routing::Route& route) {
+  double cost{0.};
+  for (int i = 0; i < route.size(); ++i) {
+    cost = std::max(ComputeCost(route.Get(i)), cost);
+  }
+  return cost;
+}
+
+// Filters routing::Routes from @p routes based on @p routing_constraints.
+// @param routing_constraints The constraints to evaluate over @p routes.
+// @param routes The collection of routing::Routes that connect the start and end.
+// @return A subset of @p routes that comply with @p routing_constraints.
```

Comment:
Consider documenting whether the relative ordering of the routes in the return value is the same as that in `@p routes`.

Reply:
Fixed in [e674557](https://github.com/maliput/maliput/pull/619/commits/e674557d805baf0219c485de6f242853285f2ca4)